### PR TITLE
Proposal to use an enum to handle Interface types

### PR DESCRIPTION
**Pull request by maackle**
* Removes the need for stringly-typed `InterfaceConfiguration` and `StorageConfiguration`, using enums instead
* Uses the `Default` impl for some struct fields instead of using `Option` to specify optional TOML fields

### CODE REVIEW

File: container_api/src/config.rs [hunk 1]
```diff
@@ -30,26 +30,31 @@ use toml;
...
+    #[serde(default)]
+    pub instances: Vec<InstanceConfiguration>,
     /// List of interfaces any UI can use to access zome functions
-    pub interfaces: Option<Vec<InterfaceConfiguration>>,
+    #[serde(default)]
+    pub interfaces: Vec<InterfaceConfiguration>,
     /// List of bridges between instances
-    pub bridges: Option<Vec<Bridge>>,
+    #[serde(default)]
+    pub bridges: Vec<Bridge>,
```
Comment (sphinxc0re) [inline container_api/src/config.rs]: All these defaults can't be right. Is an empty configuration a valid one?

File: container_api/src/config.rs [hunk 2]
```diff
@@ -177,13 +177,14 @@ pub struct LoggerConfiguration {
...
-    pub password: Option<String>,
-    pub url: Option<String>,
-    pub path: Option<String>,
+#[serde(tag="type")]
+pub enum StorageConfiguration {
+    #[serde(rename = "memory")]
+    Memory,
+    #[serde(rename = "file")]
+    File {
+        path: String,
```
Comment (sphinxc0re) [inline container_api/src/config.rs]: Please make this a `PathBuf`

File: container_api/src/config.rs (hunk 1 above)
Comment (maackle) [reply container_api/src/config.rs]: @sphinxc0re sure, it's OK to just not specify bridges in the config if you don't need them right?

It makes it so absence of a list corresponds to the empty list

Comment (maackle) [reply container_api/src/config.rs]: @sphinxc0re which would you recommend should be required? `agents` and `dna`?

Comment (sphinxc0re) [reply container_api/src/config.rs]: I'm not sure, that's why I'm asking. What exactly does the agent need to get running?

File: container_api/src/config.rs (hunk 2 above)
Comment (maackle) [reply container_api/src/config.rs]: @sphinxc0re I started doing this but if you follow the trail, this gets fed into several other functions including `create_context` which also take a `String` and should probably take a `Path`. I'd like to defer it to a separate small PR.

File: container_api/src/config.rs (hunk 1 above)
Comment (maackle) [reply container_api/src/config.rs]: @sphinxc0re you're right, according to `check_consistency`, agents, dnas, and instances should be required, so I changed it.

Comment (lucksus) [reply container_api/src/config.rs]: 👍

**Comment by zippy** [comment]
I'd like to be able to merge this but it looks like some tests are failing...

File: container_api/src/config.rs (hunk 2 above)
Comment (sphinxc0re) [reply container_api/src/config.rs]: 😱 When a path is needed `PathBuf` should be used. This sounds like a followup.

File: container_api/src/config.rs [hunk 3]
```diff
@@ -307,46 +312,63 @@ pub mod tests {
...
+            panic!("Expected `file` type StorageConfiguration!")
+        }
+
+        let interfaces = config.interfaces;
+        let interface_config_0 = interfaces.get(0).unwrap();
+        let interface_config_1 = interfaces.get(1).unwrap();
+        if let InterfaceProtocol::Websocket { port } = interface_config_0.protocol {
+            assert_eq!(port, 8888);
+        } else {
+            panic!();
```
Comment (sphinxc0re) [inline container_api/src/config.rs]: Please specify a proper error message here.

File: container_api/src/config.rs [hunk 4]
```diff
@@ -307,46 +312,63 @@ pub mod tests {
...
+        }
+        if let InterfaceProtocol::DomainSocket { ref file } = interface_config_1.protocol {
+            assert_eq!(file, "/tmp/holochain.sock");
+        } else {
+            panic!();
```
Comment (sphinxc0re) [inline container_api/src/config.rs]: See above

**Pull request merged**